### PR TITLE
core(runner): share computedCache between gather/audit

### LINKED ISSUE #11132: Make traceOfTab cacheable between gathering and auditing
Currently we have several instances of computeTraceOfTab in gathering phase that won't be cached when audits reinvoke it. This is a fairly expensive computation so we should be able to cache the result when the input doesn't change.

### PR DESCRIPTION
**Pull request by patrickhulce**
**Summary**
Hoists the `computedCache` in runner to the entire run so gatherers can use traceOfTab as well.

**Related Issues/PRs**
closes #11132 
ref #11138 
ref #11105 
ref #11094

### CODE REVIEW

File: types/gatherer.d.ts [hunk 1]
```diff
@@ -18,7 +19,8 @@ declare global {
       disableJavaScript?: boolean;
       passConfig: Config.Pass
       settings: Config.Settings;
-      options?: object;
+      computedCache: Immutable<Map<string, ArbitraryEqualityMap>>;
+      options?: Immutable<Record<string, any>>;
```
Comment (patrickhulce) [inline types/gatherer.d.ts:23]: tightened this up because `object` is pretty meaningless

**Comment by brendankenny** [comment]
I'd like to strongly consider avoiding most/all of the new interdependencies this would create and be willing to run `traceOfTab` twice, once during gathering (and include it on `traceData` as you're already doing here) and once during auditing. Sharing makes context and calling more complicated (like all the tests that need to be updated for this PR) and it makes dreams of incremental/parallel auditing require more refactoring work in the future than they would currently.

I think the execution cost is less than it appears (in fact, we're already shipping running it twice in 6.1 with the current `trace-elements`). On the (possibly underprovisioned?) HTTPArchive machines in June, the median time for `lh:computed:TraceOfTab` is 65ms, 99th percentile is 774ms, 99.9th was 1.8s.

Not that bad, and a lot of those seem to be one-off slowdowns (spikes from other processes, etc), too. Joining with the May run:
- of all the runs where `lh:computed:TraceOfTab` took longer than 1s in May, 50% of them took < 250ms in June, only about 15% of them also took greater than 1s.
- of all the runs where `lh:computed:TraceOfTab` took longer than 3s in May, 50% of them took < 682ms in June, only about 15% of them also took greater than 3s.

In addition, running some of the worst sites locally, even 13s traceOfTab time takes < 100ms, so I think there may be other VM things going on.

Here's a [list of some of the worst sites](https://docs.google.com/spreadsheets/d/1T-cm8Z10MjdiPP-DT13KIURpG6oIepm2Gjlim7nSdLU/preview) (>5s in May) for individual investigation.

In conclusion: doing it twice isn't bad, really, and keeps things much simpler :D :?

**Comment by patrickhulce** [comment]
Thanks for pulling all that data together @brendankenny! I'm shocked (and a little suspicious of our instrumentation 😅 ) that the 99th percentile has so dramatically improved without any effort on our part. I guess we did a much better job driving it down in OYB than I remembered.

I definitely should have been clearer in the meeting that I'm not concerned with the median or even really the 90th percentile case where traces are fairly small. Trace processing is just one part of LH where the cost can and does balloon for very busy sites. Doing more and more copies of something that can take ~13 seconds adds up to very user-noticeable chunks really quickly, something we don't want even in a small number of cases. I think we agree here though and now it's just a discussion of doing it twice or once.

> I'd like to strongly consider avoiding most/all of the new interdependencies this would create 

> Sharing makes context and calling more complicated (like all the tests that need to be updated for this PR) and it makes dreams of incremental/parallel auditing require more refactoring work in the future than they would currently.

Sorry, I'm still not understanding what you're referring to here. Specifically I'm not sure how this creates a new web of interdependencies beyond what invoking trace-of-tab in the gathering phase already does. I assume you're not referring to computed artifacts in general, but the fact that the same `Map` is used in gathering and auditing? If so, I'm not seeing how any of that prevents parallel auditing. 

Yes, gather-runner tests will need to be updated to pass in a computedCache (I prolly shoulda marked this as draft anticipating this conversation holding the test updates ;)), but I'm not sure updating an internal-only API with an extra `new Map()` line is a reason to subject actual users to unnecessary increased time and resources, even if a majority won't notice. I mean that's kinda our whole argument to web developers with the Lighthouse performance category 😆

File: types/gatherer.d.ts [hunk 2]
```diff
@@ -28,6 +30,7 @@ declare global {
       networkRecords: Array<Artifacts.NetworkRequest>;
       devtoolsLog: DevtoolsLog;
       trace?: Trace;
+      traceOfTab?: Artifacts.TraceOfTab;
```
Comment (brendankenny) [inline types/gatherer.d.ts:33]: What's the point of providing `computedCache` *and* `traceOfTab` to gatherers?
